### PR TITLE
Document control plane monitoring

### PR DESCRIPTION
**Pull request by serathius**
/cc @RainbowMango 
Forked from PR https://github.com/kubernetes/website/pull/17471#discussion_r346313123

Based on KEP https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/20190404-kubernetes-control-plane-metrics-stability.md#show-hidden-metrics

### CODE REVIEW

**Comment by serathius** [comment]
Thanks @sftim for your comments on previous PR, applying them here

File: content/en/docs/concepts/cluster-administration/monitoring.md [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+## Metrics in Kubernetes
+
+Kubernetes control plane components publish metrics in Prometheus text format. In most cases those metrics are available on `/metrics` endpoint of the HTTP server. For components that doesn't expose endpoint by default it can be enabled using `--bind-address` flag.
+
+Examples of those components:
+* {{< glossary_tooltip term_id="kube-controller-manager" text="kube-controller-manager" >}}
+* {{< glossary_tooltip term_id="kube-proxy" text="kube-proxy" >}}
+* {{< glossary_tooltip term_id="kube-apiserver" text="kube-apiserver" >}}
+* {{< glossary_tooltip term_id="kube-scheduler" text="kube-scheduler" >}}
+* kube-proxy
```
Comment (sftim) [inline content/en/docs/concepts/cluster-administration/monitoring.md]: Duplicated. An error I think?

File: content/en/docs/concepts/cluster-administration/monitoring.md [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+* kube-apiserver
+* kube-scheduler
+* kubelet (metrics exposed on `/metrics/cadvisor` and `/metrics/resource` do not have same guarantee)
```
Comment (sftim) [inline content/en/docs/concepts/cluster-administration/monitoring.md]:
```suggestion
*  {{< glossary_tooltip term_id="kubelet" text="kubelet" >}} (metrics exposed on `/metrics/cadvisor` and `/metrics/resource` do not have same guarantee)
```

File: content/en/docs/concepts/cluster-administration/monitoring.md [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+## Show Hidden Metrics
+
+As described above, admins can enable hidden metrics through a command-line flag on a specific binary. This intends to be used as an escape hatch for admins if they missed the migration of the metrics deprecated in the last release.
+
+The flag `show-hidden-metrics-for-version` takes a version for which you want to show metrics deprecated in that release. The version is expressed as x.y, where x is the major version, y is the minor version. The patch version is not needed even though a metrics can be deprecated in a patch release, the reason for that is the metrics deprecation policy runs against the minor release.
+
+The flag can only take the previous minor version as it's value. All metrics hidden in previous will be emitted if admins set the previous version to `show-hidden-metrics-for-version`. The too old version is not allowed because this violates the metrics deprecated policy.
+
+Take metric `A` as an example, here assumed that `A` is deprecated in 1.n. According to metrics deprecated policy, we can reach the following conclusion:
```
Comment (sftim) [inline content/en/docs/concepts/cluster-administration/monitoring.md]:
```suggestion
Take metric `A` as an example, here assumed that `A` is deprecated in  version 1.n. According to the deprecation policy for metrics:
```

File: content/en/docs/concepts/cluster-administration/monitoring.md [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+* In release `1.n`, the metric is deprecated, and it can be emitted by default.
+* In release `1.n+1`, the metric is hidden by default and it can be emitted by command line `show-hidden-metrics-for-version=1.n`.
+* In release `1.n+2`, the metric should be removed from the codebase. No escape hatch anymore.
+
+So, if admins want to enable metric `A` in release `1.n+1`, they should set `1.n` to the command line flag. That is `show-hidden-metrics=1.n`.
```
Comment (sftim) [inline content/en/docs/concepts/cluster-administration/monitoring.md]:
```suggestion
If you're using n release `1.n+1` and  want to enable metric `A`,, enable set `1.n` hidden metrics via command line: `--show-hidden-metrics=1.n`.
```

This line might be easier to explain with actual numbers (eg set `n` to _11_).

File: content/en/docs/concepts/cluster-administration/monitoring.md [hunk 5]
```diff
@@ -0,0 +1,104 @@
+---
+title: Monitoring Control Plane Components
```
Comment (sftim) [inline content/en/docs/concepts/cluster-administration/monitoring.md]: Optional change if you prefer it
```suggestion
title: Metrics For The Kubernetes Control Plane
```

File: content/en/docs/concepts/cluster-administration/monitoring.md [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+
+System component metrics can give a better look into what is happening inside them. Metrics are particularly useful for building dashboards and alerts.
+
+{{% /capture %}}
+
+{{% capture body %}}
+
+## Metrics in Kubernetes
+
+Kubernetes control plane components publish metrics in Prometheus text format. In most cases those metrics are available on `/metrics` endpoint of the HTTP server. For components that doesn't expose endpoint by default it can be enabled using `--bind-address` flag.
```
Comment (sftim) [inline content/en/docs/concepts/cluster-administration/monitoring.md]: Consider moving “Kubernetes control plane components publish metrics in Prometheus text format” into the overview, and rewording to suit.

File: content/en/docs/concepts/cluster-administration/monitoring.md [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+```
+# HELP some_counter this counts things
+# TYPE some_counter counter
+some_counter 0
+```
+
+After deprecation:
+
+```
+# HELP some_counter (Deprecated from 1.15) this counts things
```
Comment (RainbowMango) [inline content/en/docs/concepts/cluster-administration/monitoring.md]:
```suggestion
# HELP some_counter (Deprecated since 1.15.0) this counts things
```

**Comment by RainbowMango** [comment]
In general, I think we should add a feature document like this.  We can hold for a while because some stuff not done.

/milestone v1.18

**Comment by sftim** [comment]
If this pull request should not merge until v1.17 is released, I recommend adding a /hold

**Comment by RainbowMango** [comment]
thanks @sftim 

> If this pull request should not merge until v1.17 is released, I recommend adding a /hold

/hold

File: content/en/docs/concepts/cluster-administration/monitoring.md [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+
+After deprecation:
+
+```
+# HELP some_counter (Deprecated since 1.15.0) this counts things
+# TYPE some_counter counter
+some_counter 0
+```
+
+Hidden metrics will no longer be exposed by default; to use a hidden metric, you need to override the configuration for the relevant cluster component.
```
Comment (sftim) [inline content/en/docs/concepts/cluster-administration/monitoring.md]:
```suggestion
Once a metric is hidden then by default the metrics is not published for scraping. To use a hidden metric, you need to override the configuration for the relevant cluster component.
```

File: content/en/docs/concepts/cluster-administration/monitoring.md [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+
+Deleted metrics will no longer be available.
```
Comment (sftim) [inline content/en/docs/concepts/cluster-administration/monitoring.md]:
```suggestion
Once a metric is deleted, the metric is not published. You cannot change this using an override.
```

**Comment by sftim** [comment]
/unhold
because v1.17 is out
@serathius did you want to make further changes?

**Comment by serathius** [comment]
We should wait for lgtm from other reviewers
/cc @logicalhan 
/cc @brancz 
Could you take a look?

File: content/en/docs/concepts/cluster-administration/monitoring.md [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+
+Metrics in Kubernetes control plane components are exposed in Prometheus text format.
+
+{{% /capture %}}
+
+{{% capture body %}}
+
+## Metrics in Kubernetes
+
+In most cases those metrics are available on `/metrics` endpoint of the HTTP server. For components that doesn't expose endpoint by default it can be enabled using `--bind-address` flag.
```
Comment (logicalhan) [inline content/en/docs/concepts/cluster-administration/monitoring.md]: nit: remove "those"

File: content/en/docs/concepts/cluster-administration/monitoring.md [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+
+Examples of those components:
+* {{< glossary_tooltip term_id="kube-controller-manager" text="kube-controller-manager" >}}
+* {{< glossary_tooltip term_id="kube-proxy" text="kube-proxy" >}}
+* {{< glossary_tooltip term_id="kube-apiserver" text="kube-apiserver" >}}
+* {{< glossary_tooltip term_id="kube-scheduler" text="kube-scheduler" >}}
+* {{< glossary_tooltip term_id="kubelet" text="kubelet" >}}
+
+Note that {{< glossary_tooltip term_id="kubelet" text="kubelet" >}} also exposes metrics in `/metrics/cadvisor` and `/metrics/resource` endpoints. Those metrics do not have same lifecycle.
```
Comment (logicalhan) [inline content/en/docs/concepts/cluster-administration/monitoring.md]: and `/metrics/probes`

File: content/en/docs/concepts/cluster-administration/monitoring.md [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+
+Alpha metric →  Stable metric →  Deprecated metric →  Hidden metric → Deletion
+
+Alpha metrics have no stability guarantees; as such they can be modified or deleted at any time.
+
+Stable metrics can be guaranteed to not change; Specifically, stability means:
+
+* the metric itself will not be deleted (or renamed)
+* the type of metric will not be modified
+* no labels can be added or removed from this metric
```
Comment (brancz) [inline content/en/docs/concepts/cluster-administration/monitoring.md]: label values can though, I think that should be called out, also I was under the impression that last we discussed it, adding labels would be ok

File: data/concepts.yml [hunk 13]
```diff
@@ -116,6 +116,7 @@ toc:
   - docs/concepts/cluster-administration/networking.md
   - docs/concepts/cluster-administration/network-plugins.md
   - docs/concepts/cluster-administration/logging.md
+  - docs/concepts/cluster-administration/monitoring.md
```
Comment (sftim) [inline data/concepts.yml]: A new section on monitoring is a great idea.

I recommend:
- move https://kubernetes.io/docs/concepts/cluster-administration/controller-metrics/ into this section
- configure a redirect to the new URI

Also: how about adding an issue to document _workload_ monitoring? It'd be a good change that doesn't have to happen in this PR.

File: content/en/docs/concepts/cluster-administration/monitoring.md (hunk 12 above)
Comment (serathius) [reply content/en/docs/concepts/cluster-administration/monitoring.md]: This is based on KEP, if something changed we should update it

File: data/concepts.yml (hunk 13 above)
Comment (serathius) [reply data/concepts.yml]: How to configure redirect?

Comment (serathius) [reply data/concepts.yml]: Used aliases, PTAL if it make sense.

**Comment by serathius** [comment]
ping @sftim

**Comment by sftim** [comment]
Afraid I'm likely to be short on time until mid-February(!)

**Comment by serathius** [comment]
Thanks for info!
/assign @bradtopol

**Comment by bradtopol** [comment]
Assigning this to this weeks PR wrangler for shepherding.
/unassign
/assign @kbarnard10

**Pull request merged**